compiler/rustc_trait_selection/src/traits/fulfill.rs

```diff
@@ -438,29 +438,53 @@ impl<'a, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'tcx> {
-                    match ct.kind() {
-                        ty::ConstKind::Infer(ty::InferConst::Var(vid)) => {
+                    let ct_ty = match ct.kind() {
+                        ty::ConstKind::Infer(var) => {
+                            let var = match var {
+                                ty::InferConst::Var(vid) => TyOrConstInferVar::Const(vid),
+                                ty::InferConst::EffectVar(vid) => TyOrConstInferVar::Effect(vid),
+                                ty::InferConst::Fresh(_) => {
+                                    bug!("encountered fresh const in fulfill")
+                                }
+                            };
                             pending_obligation.stalled_on.clear();
-                            pending_obligation.stalled_on.extend([TyOrConstInferVar::Const(vid)]);
-                            ProcessResult::Unchanged
+                            pending_obligation.stalled_on.extend([var]);
+                            return ProcessResult::Unchanged;
                         }
                         ty::ConstKind::Error(_) => return ProcessResult::Changed(vec![]),
-                        _ => {
-                            match self.selcx.infcx.at(&obligation.cause, obligation.param_env).eq(
-                                // Only really excercised by generic_const_exprs
-                                DefineOpaqueTypes::Yes,
-                                // THISPR
-                                todo!(),
-                                ty,
-                            ) {
-                                Ok(inf_ok) => {
-                                    ProcessResult::Changed(mk_pending(inf_ok.into_obligations()))
-                                }
-                                Err(_) => ProcessResult::Error(FulfillmentErrorCode::Select(
-                                    SelectionError::Unimplemented,
-                                )),
-                            }
+                        ty::ConstKind::Value(ty, _) => ty,
+                        ty::ConstKind::Unevaluated(uv) => self
+                            .selcx
+                            .infcx
+                            .tcx
+                            .type_of(uv.def)
+                            .instantiate(self.selcx.infcx.tcx, uv.args),
+                        // FIXME(generic_const_exprs): we should construct an alias like
+                        // `<lhs_ty as Add<rhs_ty>>::Output` when this is an `Expr` representing
+                        // `lhs + rhs`.
+                        ty::ConstKind::Expr(_) => {
+                            return ProcessResult::Changed(mk_pending(vec![]));
+                        }
+                        // FIXME(BoxyUwU): this doesnt seem right
+                        ty::ConstKind::Placeholder(_) => {
+                            return ProcessResult::Changed(mk_pending(vec![]));
                         }
```

Comment:
will this just always stall? Is this even reachable? (and why don't we do the same as the new solver and look it up in the param env?

Reply:
In the new solver we canonicalize `ConstKind::Param` to `ConstKind::Placeholder` so we turn param env's with `ConstArgHasType(T/#0, usize)` into `ConstArgHasType(!0, usize)` so we will actually have candidate for placeholders. (But we still wont have any in the new solver for higher ranked consts, i.e. `for<const N: usize>`) after instantiating the binder with placeholders)

Reply:
I think also this won't stall it will just succeed (i.e. equivalent `Ok(Certainty::Yes)` in the new solver)

Reply:
So I think this is probably unreachable because the only way we should get placeholders is via instantiating a binder such as `for<const N: usize.`. Those are only writeable via `feature(non_lifetime_binders)` and that feature is super unstable (i.e. it doesn't really matter if it's a bit broken here). `compare_impl_item` artificially creates a `ParamEnv` with `for<const N: usize>` however since its  a `ParamEnv` we're only ever going to end up instantiating that binder with inference variables. (we dont seem to normalize this env afaict).

I think I should probably just turn this into an `unreachable!()` so worst case we hit an ICE here instead of unsoundness.